### PR TITLE
Sync device API to edge

### PR DESCRIPTION
**Pull request by kkBill**
**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:
sync Device API to edge, then users can get a complete Device API at edge.

**Which issue(s) this PR fixes**:

Fixes #2412

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by fisherxu** [comment]
Thanks for the contributions, still working in progress or have ready for review?

**Comment by kkBill** [comment]
> Thanks for the contributions, still working in progress or have ready for review?

yes, it's ready for review.

**Comment by fisherxu** [comment]
> > Thanks for the contributions, still working in progress or have ready for review?
> 
> yes, it's ready for review.

We need to make the ci green first. :) @kkBill

**Review by fisherxu — COMMENTED** [review]
For the upstream [here](https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/controller/upstream.go#L114), can we send the device.status from egdeside?

**Comment by daixiang0** [comment]
Please use `git rebase upstream/master`.

**Comment by kkBill** [comment]
> Please use `git rebase upstream/master`.

the e2e test failure has been fixed, but something wrong happens about the DCO. It seems that I screwed up it because I am unfamiliar with git operation. I will have a try again.

**Comment by daixiang0** [comment]
To sign DCO in your PR, you will need to use `git commit -s -m "your commit message"` to commit your change. Further info: https://github.com/apps/dco

File: edge/pkg/devicetwin/dtmanager/membership.go [hunk 1]
```diff
@@ -65,7 +66,96 @@ func initMemActionCallBack() {
 	memActionCallBack[dtcommon.MemGet] = dealMembershipGet
 	memActionCallBack[dtcommon.MemUpdated] = dealMembershipUpdate
 	memActionCallBack[dtcommon.MemDetailResult] = dealMembershipDetail
+	memActionCallBack[dtcommon.MemDeviceCRDInsert] = dealDeviceCrdInsert
+	memActionCallBack[dtcommon.MemDeviceCRDDelete] = dealDeviceCrdDelete
+	memActionCallBack[dtcommon.MemDeviceCRDUpdate] = dealDeviceCrdUpdate
 }
+
+func dealDeviceCrdInsert(context *dtcontext.DTContext, resource string, msg interface{}) (interface{}, error) {
+	klog.Infof("Membership event")
```
Comment (daixiang0) [inline edge/pkg/devicetwin/dtmanager/membership.go:75]: V(4)

File: edge/pkg/devicetwin/process.go [hunk 2]
```diff
@@ -234,7 +237,20 @@ func classifyMsg(message *dttype.DTMessage) bool {
 			}
 			message.Msg.Content = content
 		}
-		if strings.Contains(message.Msg.Router.Resource, "membership/detail") {
+		if strings.Contains(message.Msg.Router.Resource, "device_crd") { //sync Device CRD
+			operation := message.Msg.GetOperation()
+			action := ""
```
Comment (daixiang0) [inline edge/pkg/devicetwin/process.go:242]: Assign message.Action directly

File: edge/pkg/devicetwin/dtmanager/membership.go (hunk 1 above)
Comment (kkBill) [reply edge/pkg/devicetwin/dtmanager/membership.go:75]: ok, got it

File: edge/pkg/devicetwin/process.go (hunk 2 above)
Comment (kkBill) [reply edge/pkg/devicetwin/process.go:242]: ok, got it

**Comment by kkBill** [comment]
> For the upstream [here](https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/controller/upstream.go#L114), can we send the device.status from egdeside?

For now, it is not supported yet, I just send a complete Device API to edge and store it 😄 . For upstream you are mentioned, it's counterpart at edge is [here](https://github.com/kubeedge/kubeedge/blob/master/edge/pkg/devicetwin/dtmanager/twin.go#L113). The logic about device at edge is strongly correlated with [DTContext](https://github.com/kubeedge/kubeedge/blob/master/edge/pkg/devicetwin/dtcontext/dtcontext.go#L20), which plays an important role and used for 4 sub-modules of `devicetwin`, now it stores device(not Device API) in `DTContext.DeviceList`. In my opinion, it is hard to modify one particular sub-module. If we want to make Device API available at edge, I think we need to store Device API in DTContext first then refactor other components.

File: edge/pkg/devicetwin/dtmanager/membership.go (hunk 1 above)
Comment (daixiang0) [reply edge/pkg/devicetwin/dtmanager/membership.go:75]: Unify them, now there is not a common rule for log level.

**Pull request closed without merging**